de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/concept/types/structural/level/HierarchyLevelChecker.java

```diff
@@ -0,0 +1,368 @@
+			IBeanStructuralElementInstance elementToCheck) {
+		IHierarchyLevel childLevel = getLevelOfBean(child);
+		int treeDistance = getTreeDistance(child, elementToCheck);
+		int minLevel = levels.indexOf(childLevel) - treeDistance;
+
+		// Go through all the levels between the child level and the one with a
+		// distance of the tree depth and check if their optional. If so, then the
+		// minimum applicable level index decreases as we can also skip the level
+		for (int i = levels.indexOf(childLevel) - 1; i >= 0 && i >= levels.indexOf(childLevel) - treeDistance; i--) {
+			if (levels.get(i).isOptional()) {
+				minLevel--;
+			}
+		}
+
+		return minLevel;
+	}
+
+	/**
+	 * Get the maximum level index considering the tree distance to the next parent
+	 * element with level
+	 * 
+	 * @param parent
+	 *            the next parent element with level
+	 * @param elementToCheck
+	 *            the element to check
+	 * @return the maximum level index
+	 */
+	private int getMaxIndexFromTreeDistanceOfParent(IBeanStructuralElementInstance parent,
+			IBeanStructuralElementInstance elementToCheck) {
+		IHierarchyLevel parentLevel = getLevelOfBean(parent);
+		int treeDistance = getTreeDistance(elementToCheck, parent);
+		int maxLevel = levels.indexOf(parentLevel) + treeDistance;
+
+		// Go through all the levels between the parent level and the one with a
+		// distance of the tree depth and check if their optional. If so, then the
+		// maximum applicable level index increases as we can also skip the level
+		for (int i = levels.indexOf(parentLevel) + 1; i <= levels.indexOf(parentLevel) + treeDistance
+				&& i < levels.size(); i++) {
+			if (levels.get(i).isOptional()) {
+				maxLevel++;
+			}
+		}
+
+		return maxLevel;
+	}
+
+	/**
+	 * Get the level an arbitrary bean is on
+	 * 
+	 * @param bean
+	 *            the bean to get level from
+	 * @return the level
+	 */
+	private IHierarchyLevel getLevelOfBean(IBeanStructuralElementInstance bean) {
+		List<IHierarchyLevel> foundLevels = getLevelsOfBean(bean);
+		if (foundLevels.size() > 1) {
+			throw new IllegalArgumentException("Element " + bean + " is on multiple levels: " + foundLevels);
+		}
+		if (foundLevels.size() == 1) {
+			return foundLevels.get(0);
+		}
+		return null;
+	}
+
+	/**
+	 * Get all levels a bean is on
+	 * @param bean 
+	 * @return list of levels
+	 */
+	private List<IHierarchyLevel> getLevelsOfBean(IBeanStructuralElementInstance bean) {
+		return levels.stream()
+				.filter(level -> level.isOnLevel(bean))
+				.collect(Collectors.toList());
+	}
+	
+	/**
+	 * Get the first parent with level or null if not existing
+	 * 
+	 * @param bean
+	 *            the SEIBean
+	 * @return the next parent with level
+	 */
+	private IBeanStructuralElementInstance getFirstParentWithLevel(IBeanStructuralElementInstance bean) {
+		IBeanStructuralElementInstance parent = bean.getParentSeiBean();
+
+		if (parent != null) {

```

Comment:
Maybe a 

`if (parent == null) {
     return null;
}`

in the beginning to reduce nesting.